amqp/src/test/scala/akka/stream/alpakka/amqp/scaladsl/AmqpGraphStageLogicConnectionShutdownSpec.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.amqp.scaladsl
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.actor.ActorSystem
+import akka.dispatch.ExecutionContexts
+import akka.stream.ActorMaterializer
+import akka.stream.alpakka.amqp.{
+  AmqpCachedConnectionProvider,
+  AmqpConnectionFactoryConnectionProvider,
+  AmqpProxyConnection,
+  AmqpSinkSettings,
+  DelegatingConnectionProvider,
+  QueueDeclaration
+}
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import com.rabbitmq.client.{Connection, ConnectionFactory, ShutdownListener}
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.{Matchers, WordSpec}
+
+import scala.concurrent.{Future, Promise}
+import scala.concurrent.duration._
+
+/**
+ * see [[https://github.com/akka/alpakka/issues/883]] and
+ * [[https://github.com/akka/alpakka/pull/887]]
+ */
+class AmqpGraphStageLogicConnectionShutdownSpec extends WordSpec with Matchers with ScalaFutures {
+
+  override implicit val patienceConfig = PatienceConfig(10.seconds)
+  private implicit val executionContext = ExecutionContexts.sameThreadExecutionContext
+
+  "registers and unregisters a single connection shutdown hook per graph" in {
+    // actor system is within this test as it has to be shut down in order
+    // to verify graph stage termination
+    implicit val system = ActorSystem(this.getClass.getSimpleName + System.currentTimeMillis())
+    implicit val materializer = ActorMaterializer()
+
+    val spyConnectionPromise = Promise[Connection]
+    val shutdownsAdded = new AtomicInteger()
+    val shutdownsRemoved = new AtomicInteger()
+
+    def spy(conn: Connection): Connection = new AmqpProxyConnection(conn) {
```

Comment:
I don't think that you need the `AmqpProxyConnection` class.
Following up on my previous comment I think that you could just do:
```
    val factory = new ConnectionFactory {
      override def newConnection(): Connection = new Connection {
        override def addShutdownListener(shutdownListener: ShutdownListener): Unit = {
          shutdownsAdded.incrementAndGet()
          super.addShutdownListener(shutdownListener)
        }

        override def removeShutdownListener(shutdownListener: ShutdownListener): Unit = {
          shutdownsRemoved.incrementAndGet()
          super.removeShutdownListener(shutdownListener)
        }
      }
    }
```
Right?

Reply:
Hmm, I'm not sure this will work. I think in my current case I need a valid amqp connection, and it's not necessarily the case that extending a default, no-args construction of `Connection` is the safe action. Looking at the default implementation of `ConnectionFactory`, various properties are set, including the connection name, executor pool, etc. What I really want is a wrapper around some Connection created by the underlying factory, decorated to track shutdown listeners. (Note how much actually goes into creating a connection: https://github.com/rabbitmq/rabbitmq-java-client/blob/master/src/main/java/com/rabbitmq/client/ConnectionFactory.java#L978)

I like the idea of using a spy or a mixin, but:
- Mockito's `spy` does not claim to be thread-safe
- I can't dynamically add a mixin to a class I'm not instantiating

So, I'm not entirely sure I can get rid of the `AmqpProxyConnection`